### PR TITLE
IPv6 and IPAM support

### PR DESCRIPTION
**Pull request by Nuru**
## what
- Adds support for IPv6 configuration
- Adds preliminary support for [IP Address Manager](https://docs.aws.amazon.com/vpc/latest/ipam/what-it-is-ipam.html) (IPAM)
- Updates framework to current

## why
- Support new features by request

## migration notes

These changes should be backward compatible. However, 3 variables were deprecated and replaced, and use of the new variables is encouraged.

#### `ipv6_enabled` deprecated in favor of `assign_generated_ipv6_cidr_block`

In PR #99 released in version v0.28.0, we deprecated `assign_generated_ipv6_cidr_block` in favor of `ipv6_enabled` as part of a more general renaming of inputs to follow Cloud Posse's convention of naming boolean flags with `_enabled` at the end. However, now that we have IPAM, the name `ipv6_enabled` is misleading: it only means  `assign_generated_ipv6_cidr_block` and you have to set that to `false` in order to assign an IPv6 address from IPAM. We do not want require you to set `ipv6_enabled = false` in order to use IPv6 IPAM, so we are reversing course on this variable, deprecating `ipv6_enabled` and reviving `assign_generated_ipv6_cidr_block`.

#### `cidr_block` deprecated in favor of `ipv4_primary_cidr_block`

Now that a VPC can have multiple IPv4 CIDR blocks along with multiple IPv6 CIDR blocks, the name `cidr_block` is just too vague. Use `ipv4_primary_cidr_block` instead to set the primary IPv4 CIDR block for the VPC. 

#### `additional_cidr_blocks` deprecated in favor of `ipv4_additional_cidr_block_associations`

As with `cidr_block`, `additional_cidr_blocks` is now too vague given that the module supports both IPv4 and IPv6 CIDR blocks. Also, you now have the option of having these CIDR blocks allocated by IPAM rather than being specified directly, requiring additional input parameters. So we created the new `ipv4_additional_cidr_block_associations` and the parallel `ipv6_additional_cidr_block_associations` to replace `additional_cidr_blocks` (which, while deprecated, remains supported for now). 

Note that the new variables take maps of configurations rather than lists of configurations. This is because each configuration is deployed as a separate Terraform resource, and the resources are not easily deleted and recreated. If you supply a list and then later remove the first element of the list, Terraform will deleted and recreate all the resources because Terraform tracks them by their list index and does not detect that they have simply moved. It sees that configuration 1 has changed (it is now what used to be configuration 2) so it deletes the old 1 and creates the old 2 as the new 1. This annoying at best, and with CIDR blocks being deleted and recreated it will likely fail at worst. 

By providing the configurations as maps, you give each configuration a stable label, so that when you make a change, only the affected configuration changes. Removing a configuration does not affect the other configurations. The labels you assign are arbitrary and for your own benefit; the only requirement is that they be known at "plan time", meaning they cannot be generated by the Terraform code itself. 

## notes

Due to the difficulty of configuring and testing the AWS VPC IP Address Manager, this feature has not been tested and support for it should be considered preliminary. Please report issues as you discover them.

Some inputs have been deprecated. New inputs typically take a map rather than a list so that we can use `for_each` (for stability as the set of inputs changes) without relying on the values in the list (which can cause Terraform to complain that the results depend on values not available at "plan" time).

Supersedes and closes #103

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -1,6 +1,14 @@
 locals {
   enabled                                   = module.this.enabled
   ipv6_egress_only_internet_gateway_enabled = local.enabled && var.ipv6_egress_only_internet_gateway_enabled
+  additional_cidr_blocks_map = local.enabled ? { for v in var.additional_cidr_blocks : v => {
+    ipv4_cidr_block     = v
+    ipv4_ipam_pool_id   = null
+    ipv4_netmask_length = null
+  } } : {}
```
Comment (nitrocode) [inline main.tf:8]: Just for readability

```suggestion
  additional_cidr_blocks_map = local.enabled ? {
    for cidr_block in var.additional_cidr_blocks :
    cidr_block => {
      ipv4_cidr_block     = cidr_block
      ipv4_ipam_pool_id   = null
      ipv4_netmask_length = null
    }
  } : {}
```

File: main.tf [hunk 2]
```diff
@@ -14,15 +22,24 @@ resource "aws_vpc" "default" {
...
+  cidr_block          = local.ipv4_primary_cidr_block
+  ipv4_ipam_pool_id   = try(var.ipv4_primary_cidr_block_association.ipv4_ipam_pool_id, null)
+  ipv4_netmask_length = try(var.ipv4_primary_cidr_block_association.ipv4_netmask_length, null)
+
+  ipv6_cidr_block     = try(var.ipv6_primary_cidr_block_association.ipv6_cidr_block, null)
+  ipv6_ipam_pool_id   = try(var.ipv6_primary_cidr_block_association.ipv6_ipam_pool_id, null)
+  ipv6_netmask_length = try(var.ipv6_primary_cidr_block_association.ipv6_netmask_length, null)
+
+
+  # Additional ipv4 CIDRs are handled by aws_vpc_ipv4_cidr_block_association below
```
Comment (nitrocode) [inline main.tf:34]:
```suggestion
```

File: main.tf [hunk 3]
```diff
@@ -14,15 +22,24 @@ resource "aws_vpc" "default" {
...
+  # Additional ipv6 CIDRs are handled by aws_vpc_ipv6_cidr_block_association below
```
Comment (nitrocode) [inline main.tf:35]:
```suggestion
  # Additional ipv4 and ipv6 CIDRs are handled by aws_vpc_ipv6_cidr_block_association below
```

Comment (Nuru) [reply main.tf:35]: But that is wrong

File: outputs.tf [hunk 4]
```diff
@@ -63,3 +58,35 @@ output "additional_cidr_blocks_to_association_ids" {
...
+    i.ipv6_cidr_block
+  ]
+}
+
+output "additional_ipv6_cidr_blocks_to_association_ids" {
+  description = "A map of the additional IPv4 CIDR blocks to VPC CIDR association IDs"
+  value = {
+    for i in aws_vpc_ipv6_cidr_block_association.default :
+    i.ipv6_cidr_block => i.id
+  }
```
Comment (nitrocode) [inline outputs.tf:75]:
```suggestion
  value = locals.enabled ? {
    for i in aws_vpc_ipv6_cidr_block_association.default :
    i.ipv6_cidr_block => i.id
  } : null
```

File: outputs.tf [hunk 5]
```diff
@@ -63,3 +58,35 @@ output "additional_cidr_blocks_to_association_ids" {
     i.cidr_block => i.id
   }
 }
+
+output "additional_ipv6_cidr_blocks" {
+  description = "A list of the additional IPv4 CIDR blocks associated with the VPC"
+  value = [
+    for i in aws_vpc_ipv6_cidr_block_association.default :
+    i.ipv6_cidr_block
+  ]
```
Comment (nitrocode) [inline outputs.tf:67]:
```suggestion
  value = local.enabled ? [
    for i in aws_vpc_ipv6_cidr_block_association.default :
    i.ipv6_cidr_block
  ] : null
```

**Comment by nitrocode** [comment]
Looks pretty good.

* It would be nice to see a migration doc. It looks backwards compatible but it would nice to see a migration doc to get off the deprecated inputs.
* Does the README.yaml require any updates to show how to enable both ipv4 and ipv6
* Could you add a `TestExamplesCompleteDisabled` to ensure `enabled = false` will disable everything
    * I've been using this one its respective common.go https://github.com/cloudposse/terraform-aws-lambda-function/blob/cb734da0a221f49477b95cc80c598bb7e97f9a83/test/src/examples_complete_test.go#L59-L62

File: outputs.tf (hunk 4 above)
Comment (Nuru) [reply outputs.tf:75]: It's not necessary to null the outputs when the module is disabled. Most of this module's outputs are `string` and are `""` when disabled. For `list` and `map` it is acceptable that the outputs be empty lists and maps, and you can argue it is even preferable. At least for this module it is consistent with the older outputs.

**Comment by Nuru** [comment]
> Looks pretty good.
> 
>     * It would be nice to see a migration doc. It looks backwards compatible but it would nice to see a migration doc to get off the deprecated inputs.

I will add the notes on changes. 

> 
>     * Does the README.yaml require any updates to show how to enable both ipv4 and ipv6

Right, I forgot to update the README

> 
>     * Could you add a `TestExamplesCompleteDisabled` to ensure `enabled = false` will disable everything
>       

The `examples/complete` and `examples/deprecated` both verify that the disabled modules work without causing problems. That's generally been the extent of our testing. I do not like [your recommended solution](https://github.com/cloudposse/terraform-aws-lambda-function/blob/cb734da0a221f49477b95cc80c598bb7e97f9a83/test/src/common.go#L35) because I am OK with disabled modules outputting empty lists and maps. I'll ask for other opinions via Slack.

**Review by nitrocode — APPROVED** [review]

**Pull request merged**